SimPEG/ObjectiveFunction.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import numpy as np
+import scipy.sparse as sp
+from six import integer_types
+import warnings
+
+from . import Utils
+from .Tests import checkDerivative
+from . import Maps
+from . import Props
+
+__all__ = [
+    'BaseObjectiveFunction', 'ComboObjectiveFunction', 'L2ObjectiveFunction'
+]
+
+
+class BaseObjectiveFunction(Props.BaseSimPEG):
+
+    counter = None
+    debug = False
+
+    def __init__(self, nP=None, **kwargs):
+        self._nP = nP
+        Utils.setKwargs(self, **kwargs)
+
+    def __call__(self, x, **kwargs):
+        return self._eval(x, **kwargs)
+
+    @property
+    def nP(self):
+        if getattr(self, '_nP', None) is None:
+            self._nP = '*'
+        return self._nP
+
+    @Utils.timeIt
+    def _eval(self, x, **kwargs):
```

Comment:
My only thought with keeping `_eval` is that it can still be found with tab complete. I could be convinced either way though